modules/prometheus/src/main/scala/com/evolutiongaming/smetrics/Prometheus.scala

```diff
@@ -13,19 +13,32 @@ trait Prometheus[F[_]] {
   def write004: F[String]
 }
 
-object Prometheus {
+object Prometheus { prometheus =>
 
   def apply[F[_] : Sync](collectorRegistry: P.CollectorRegistry): Prometheus[F] =
     new Prometheus[F] {
 
       override val registry: CollectorRegistry[F] = CollectorRegistryPrometheus(collectorRegistry)
 
-      override val write004: F[String] = Sync[F].delay {
-        val writer = new StringWriter
-        TextFormat.write004(writer, collectorRegistry.metricFamilySamples)
-        writer.toString
-      }
+      override val write004: F[String] = prometheus.write004[F](collectorRegistry)
+    }
+
+  def cached[F[_] : Async](javaRegistry: P.CollectorRegistry): F[Prometheus[F]] =
+    for {
+      cachedRegistry <- CollectorRegistryPrometheus(javaRegistry).withCaching
+    } yield new Prometheus[F] {
+
+      override val registry: CollectorRegistry[F] = cachedRegistry
+
+      override val write004: F[String] = prometheus.write004[F](javaRegistry)
     }
 
   def default[F[_] : Sync]: Prometheus[F] = apply(P.CollectorRegistry.defaultRegistry)
+
+  private def write004[F[_] : Sync](registry: P.CollectorRegistry): F[String] = Sync[F].delay {
```

Comment:
I think this you should revert and make sure that `cached` is an extension method over `Prometheus`, wdyt?

Reply:
yep, redone as extension method `.withCaching`